### PR TITLE
Add configuration flag to set if RecursionDesired should be set on health checkers in Forward-plugin

### PR DESCRIPTION
**Pull request by ctryti**
### 1. Why is this pull request needed and what does it do?
We would like to configure the health checker for the proxies whether they should do recursive queries (default) or turn it off by setting health_check_non_recursive in the configuration.

### 2. Which issues (if any) are related?
Issues in our environment involve upstream queries to time out before the answer gets back because of how the dns servers (upstream) are configured. They are out of our control and we only care if they are alive.

### 3. Which documentation changes (if any) need to be made?
Configuration file for the forward plugin should reflect the new optional flag to turn off recursion on the health checker.

### 4. Does this introduce a backward incompatible change or deprecation?
No. The default is to keep recursion requests on by default as is coded in previous versions of forward plugin.

### CODE REVIEW

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "[coredns/coredns] Make health-check..." ]
>&lt;!--
>Thank you for contributing to CoreDNS!
>Please provide the following information to help us make the most of your pull request:
>--&gt;
>
>### 1. Why is this pull request needed and what does it do?
>The `forward`-plugin states in comments that it performs health-checks on upstream dns-servers by sending non-recursive `. IN NS` questions. Currently in master, it sends recursive requests.
>(https://github.com/coredns/coredns/blob/master/plugin/forward/health.go#L43)
>
>This may not be a problem in environments where dns is set up ideally. But on our premise, these recursive requests results in timeouts / 10 second late responses, which causes Coredns to think the upstreams are always unhealthy.
>
>We found this out when we took down 1 of 2 upstreams for maintainence, and we suddenly saw a high amount of failed dns-queries.

I've seen the opposite on my home router. Having recursion on is more standard than having
if off

What's your upstream running, because it sounds works than CPE

**Comment by ctryti** [comment]
We're in an air-gapped environment. Our upstreams are Active Directory DNS, which we suspect it might not be configured ideally, and that it's asking for records out on the internet which gets blocked by our firewall.

Sorry, i'm unfamiliar with the term; what does CPE mean?

We tested the change in this PR on our kubernetes-cluster, and it works as we expect. 
So, if making the health-check do non-recursive queries default is a no-go, how about making it configurable?

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Make health-c..." ]
>So, if making the health-check do non-recursive queries default is a no-go, how about making it configurable?

sadly, that is the only option, yes.

**Comment by ctryti** [comment]
Updated title and PR-description after feedback from reviewer.

File: plugin/forward/README.md [hunk 1]
```diff
@@ -52,6 +52,7 @@ forward FROM TO... {
     tls_servername NAME
     policy random|round_robin|sequential
     health_check DURATION
+    health_check_non_recursive
```
Comment (miekg) [inline plugin/forward/README.md]: It's better/cleaner to make this an option property of `health_check`, probably something like
`healthcheck DURATION [no_rec]` or something

File: plugin/forward/forward.go [hunk 2]
```diff
@@ -52,7 +52,7 @@ type Forward struct {
 
 // New returns a new Forward.
 func New() *Forward {
-	f := &Forward{maxfails: 2, tlsConfig: new(tls.Config), expire: defaultExpire, p: new(policy.Random), from: ".", hcInterval: hcInterval}
+	f := &Forward{maxfails: 2, tlsConfig: new(tls.Config), expire: defaultExpire, p: new(policy.Random), from: ".", hcInterval: hcInterval, opts: options{false, false, true}}
```
Comment (miekg) [inline plugin/forward/forward.go]: the options struct should use named paramaeters: `forceTCP: false` etc.

File: plugin/forward/health.go [hunk 3]
```diff
@@ -14,10 +14,14 @@ import (
...
 
 // dnsHc is a health checker for a DNS endpoint (DNS, and DoT).
-type dnsHc struct{ c *dns.Client }
+type dnsHc struct {
+	c                *dns.Client
+	recursionDesired bool
+}
 
 // NewHealthChecker returns a new HealthChecker based on transport.
 func NewHealthChecker(trans string) HealthChecker {
```
Comment (miekg) [inline plugin/forward/health.go]: this should probably receive a boolean or an `options` to set the recursion field

File: plugin/forward/health.go [hunk 4]
```diff
@@ -40,7 +44,11 @@ func (h *dnsHc) SetTLSConfig(cfg *tls.Config) {
 	h.c.TLSConfig = cfg
 }
 
-// For HC we send to . IN NS +norec message to the upstream. Dial timeouts and empty
+func (h *dnsHc) SetRecursionDesired(recursionDesired bool) {
```
Comment (miekg) [inline plugin/forward/health.go:48]: probably not needed.

File: plugin/forward/health_test.go [hunk 5]
```diff
@@ -43,6 +43,37 @@ func TestHealth(t *testing.T) {
 	}
 }
 
+func TestHealthNoRecursion(t *testing.T) {
```
Comment (miekg) [inline plugin/forward/health_test.go]: this doesn't test the functionality you've added. The only thing we care about is if the incoming query now has the recursion bit set or not

File: plugin/forward/health_test.go [hunk 6]
```diff
@@ -83,6 +114,47 @@ func TestHealthTimeout(t *testing.T) {
 	}
 }
 
+func TestHealthTimeoutNoRecursion(t *testing.T) {
```
Comment (miekg) [inline plugin/forward/health_test.go]: this is duplicating other tests

File: plugin/forward/setup_test.go [hunk 7]
```diff
@@ -21,21 +21,24 @@ func TestSetup(t *testing.T) {
...
-		{"forward . 127.0.0.1 {\nexcept miek.nl\n}\n", false, ".", nil, 2, options{}, ""},
-		{"forward . 127.0.0.1 {\nmax_fails 3\n}\n", false, ".", nil, 3, options{}, ""},
-		{"forward . 127.0.0.1 {\nforce_tcp\n}\n", false, ".", nil, 2, options{forceTCP: true}, ""},
-		{"forward . 127.0.0.1 {\nprefer_udp\n}\n", false, ".", nil, 2, options{preferUDP: true}, ""},
-		{"forward . 127.0.0.1 {\nforce_tcp\nprefer_udp\n}\n", false, ".", nil, 2, options{preferUDP: true, forceTCP: true}, ""},
-		{"forward . 127.0.0.1:53", false, ".", nil, 2, options{}, ""},
-		{"forward . 127.0.0.1:8080", false, ".", nil, 2, options{}, ""},
-		{"forward . [::1]:53", false, ".", nil, 2, options{}, ""},
-		{"forward . [2003::1]:53", false, ".", nil, 2, options{}, ""},
+		{"forward . 127.0.0.1", false, ".", nil, 2, options{hcRecursionDesired: true}, ""},
```
Comment (miekg) [inline plugin/forward/setup_test.go:24]: please add a single and small, contained test to test for this new property

File: plugin/forward/health_test.go (hunk 6 above)
Comment (ctryti) [reply plugin/forward/health_test.go]: Redid both tests i added. Now checks the RD-flag in the dns-requests.

File: plugin/forward/health.go (hunk 4 above)
Comment (ctryti) [reply plugin/forward/health.go:48]: Removed

File: plugin/forward/health.go (hunk 3 above)
Comment (ctryti) [reply plugin/forward/health.go]: I added a bool to HealthCheckers signature. Are you ok with this being set to true default in NewProxy()? Or should I also edit the signature of NewProxy() to include this config?

File: plugin/forward/forward.go (hunk 2 above)
Comment (ctryti) [reply plugin/forward/forward.go]: Done

File: plugin/forward/README.md (hunk 1 above)
Comment (ctryti) [reply plugin/forward/README.md]: Done. `health_check` now supports an optional `no_rec` argument.

File: plugin/forward/health_test.go (hunk 5 above)
Comment (ctryti) [reply plugin/forward/health_test.go]: Redid both tests i added. Now checks the RD-flag in the dns-requests.

File: plugin/forward/setup_test.go (hunk 7 above)
Comment (ctryti) [reply plugin/forward/setup_test.go:24]: Added a new test testing the health_check properties.

File: plugin/forward/README.md [hunk 8]
```diff
@@ -85,7 +85,10 @@ forward FROM TO... {
   * `random` is a policy that implements random upstream selection.
   * `round_robin` is a policy that selects hosts based on round robin ordering.
   * `sequential` is a policy that selects hosts based on sequential ordering.
-* `health_check`, use a different **DURATION** for health checking, the default duration is 0.5s.
+* `health_check` configure the behaviour of health checking of the upstream servers
```
Comment (miekg) [inline plugin/forward/README.md:88]: Please keep the current text and just add your new property in a new sentence after it, i.e. "The optional `no_rec` argument...`

File: plugin/forward/health.go (hunk 4 above)
Comment (miekg) [reply plugin/forward/health.go:48]: either github is not working, or you have pushed your code?

File: plugin/forward/health.go [hunk 9]
```diff
@@ -40,7 +45,14 @@ func (h *dnsHc) SetTLSConfig(cfg *tls.Config) {
...
+	h.recursionDesired = recursionDesired
+}
+func (h *dnsHc) GetRecursionDesired() bool {
```
Comment (miekg) [inline plugin/forward/health.go:51]: Drop Get from this - do we really need it btw?

File: plugin/forward/health_test.go [hunk 10]
```diff
@@ -14,10 +14,15 @@ import (
 )
 
 func TestHealth(t *testing.T) {
```
Comment (miekg) [inline plugin/forward/health_test.go:16]: don't change this function for this PR

File: plugin/forward/health_test.go [hunk 11]
```diff
@@ -39,7 +44,43 @@ func TestHealth(t *testing.T) {
...
+func TestHealthNoRecursion(t *testing.T) {
+	const expected = 1
+	i := uint32(0)
+	q := uint32(0)
+	s := dnstest.NewServer(func(w dns.ResponseWriter, r *dns.Msg) {
+		if atomic.LoadUint32(&q) == 0 { //drop the first query to trigger health-checking
+			atomic.AddUint32(&q, 1)
+			return
+		}
+		if r.Question[0].Name == "." && r.RecursionDesired == false {
```
Comment (miekg) [inline plugin/forward/health_test.go:60]: Do a t.Fatal here if recursion is set

File: plugin/forward/health_test.go [hunk 12]
```diff
@@ -39,7 +44,43 @@ func TestHealth(t *testing.T) {
 	time.Sleep(1 * time.Second)
 	i1 := atomic.LoadUint32(&i)
 	if i1 != expected {
-		t.Errorf("Expected number of health checks to be %d, got %d", expected, i1)
+		t.Errorf("Expected number of health checks with RecursionDesired==true to be %d, got %d", expected, i1)
+	}
+}
+
+func TestHealthNoRecursion(t *testing.T) {
+	const expected = 1
```
Comment (miekg) [inline plugin/forward/health_test.go:52]: We only care that recursion is unset. The other tests takes care of counting. Please minimize what you're trying to test here.

File: plugin/forward/setup_test.go (hunk 7 above)
Comment (miekg) [reply plugin/forward/setup_test.go:24]: have you pushed your updated code?

Comment (ctryti) [reply plugin/forward/setup_test.go:24]: Yeah, it starts at line 215

File: plugin/forward/health_test.go (hunk 12 above)
Comment (ctryti) [reply plugin/forward/health_test.go:52]: I can change the `expected` to a bool, but i still have to count. At least fail one query for the healthchecker to kick in.

File: plugin/forward/health.go (hunk 9 above)
Comment (ctryti) [reply plugin/forward/health.go:51]: It does feel a bity java'y. But i don't know how else to get the property out of the dnsHc-instance. Any tips?

File: plugin/forward/health.go (hunk 4 above)
Comment (ctryti) [reply plugin/forward/health.go:48]: Sorry, I thought i removed this function, but it was a similar one from another file. I'm not sure how else to bring the boolean from the setup-function over to the dnsHc without changing a couple of function-signatures.

Comment (miekg) [reply plugin/forward/health.go:48]: yeah, looking back at this code - it looks a bit over-engineered and needs some drastic simplications.

**Review by miekg — APPROVED** [review]
as said in comment, we probably want to revisit this entire setup, looking back feel super generic for no good reasons.

**Pull request merged**